### ISSUE TITLE
OOM downloading new mail, if the local mailbox is already too big

### ISSUE THREAD

**Issue by stevecotton** [k9mail/k-9#2061]
I have many messages with attachments that I've already downloaded to my phone, after deleting some messages and compacting the account the app is still using 89MB of storage.  The 5.0 branch coped with this, but after upgrading to 5.202 I get an OOM when trying to retrieve more mail.

This is on a POP3 account, with the phone leaving the email on the server. However, a separate computer is flushing the messages when it downloads them, there aren't many messages on the server itself.

### Expected behavior
The app copes with the unreasonably large amount of local data.

### Actual behavior
After pulling to refresh, the background service notification shows a spinner for about a minute, with no data transferring over the network. ADB shows lots of garbage collector activity, until:
```
D/dalvikvm(14472): GC_FOR_ALLOC freed 4K, 5% free 62794K/65536K, paused 203ms, total 203ms
I/dalvikvm-heap(14472): Forcing collection of SoftReferences for 16400-byte allocation
I/dalvikvm-heap(14472): Clamp target GC heap from 70.648MB to 64.000MB
D/dalvikvm(14472): GC_BEFORE_OOM freed <1K, 5% free 62793K/65536K, paused 218ms, total 218ms
E/dalvikvm-heap(14472): Out of memory on a 16400-byte allocation.
I/dalvikvm(14472): "MessagingController" prio=5 tid=11 RUNNABLE
I/dalvikvm(14472):   | group="main" sCount=0 dsCount=0 obj=0x42741638 self=0x57a71400
I/dalvikvm(14472):   | sysTid=14499 nice=10 sched=0/0 cgrp=apps/bg_non_interactive handle=1470566272
I/dalvikvm(14472):   | state=R schedstat=( 79651612389 22664743317 10558 ) utm=7758 stm=206 core=0
I/dalvikvm(14472):   at java.util.HashMap.makeTable(HashMap.java:~569)
I/dalvikvm(14472):   at java.util.HashMap.doubleCapacity(HashMap.java:589)
I/dalvikvm(14472):   at java.util.HashMap.put(HashMap.java:419)
I/dalvikvm(14472):   at com.fsck.k9.controller.MessagingController.synchronizeMailboxSynchronous(MessagingController.java:802)
I/dalvikvm(14472):   at com.fsck.k9.controller.MessagingController$36.run(MessagingController.java:4010)
I/dalvikvm(14472):   at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:204)
I/dalvikvm(14472):   at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:115)
I/dalvikvm(14472):   at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:173)
I/dalvikvm(14472):   at java.lang.Thread.run(Thread.java:841)
```

IIUC, MessagingCortroller.java:802 is putting the contents of all local messages in to a HashMap:
```
            List<? extends Message> localMessages = localFolder.getMessages(null);
            Map<String, Message> localUidMap = new HashMap<>();
            for (Message message : localMessages) {
                localUidMap.put(message.getUid(), message);
            }
```

I deleted some of the local messages, emptied the trash and compacted the account.  K-9 then reached MessagingController.java:905 before crashing. Looking at the already-logged bugs #810 and #1241, both occur later in that function, but it seems that the localUidMap would still be in scope and holding on to the memory when both of those two crashes happen.  I think I just have an even more unreasonably large local mailbox than them.

### Steps to reproduce
1. Have a large inbox, including people sending scanned documents
2. Try to download more.

### Environment
K-9 Mail version: 5.202, Fdroid version

Android version: 4.3

Account type (IMAP, POP3, WebDAV/Exchange): POP3

**Comment by philipwhiuk** [comment]
I would guess that the best way to fix this is `WeakReference`'ing some of the contents of Message. But we really need to first check that it is `Message` that's huge here and not something else.

Certainly attachments ought to be irrelevant here - we shouldn't have them in scope once we've stored them.

**Comment by cketti** [comment]
The proper fix is not trying to load information about all messages into memory at once. Synchronizing in batches makes things certainly a bit more complicated. But we should be able to get rid of all kinds of OOM errors.

**Comment by Valodim** [comment]
I don't think `WeakReferences` are a very good idea, at least not as a first iteration. It moves the problems around in ways that are very difficult to reason about.

**Comment by Valodim** [comment]
For the particular case of `MessagingController:802`, the call to `getMessages` is super heavy - it fills a hashmap, parsing the headers of all messages in the process through `populateFromGetMessageCursor`. What we actually use are only the message uid plus the `olderThan` method, which accesses the internal and sent date of that message.

The solution to this issue here is to simply optimize that call away. I'll give it a shot.

**Comment by philipwhiuk** [comment]
Sounds good Valodim :)

**Comment by Valodim** [comment]
fixed with #2069 and merged in 5.2-MAINT
